### PR TITLE
[Improve](InPredict) enhance in predict with array type

### LINKED ISSUE #31157: [Feature] Support Array Type In Function and equal function
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

support two opeartor of array function
1. select [1,2] in ([1,2],[3,4])
2. select [1,2] = [1,1];

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by taptao**
## Proposed changes

Issue Number: close https://github.com/apache/doris/issues/31157

Support in predict with array type.
Example：select [1,2] in ([1,2],[3,4])

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by taptao** [comment]
run buildall

File: be/src/vec/functions/function_array_in.h [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+namespace doris {
+namespace vectorized {
+struct ColumnRowRef;
+}
+} // namespace doris
+
+namespace doris::vectorized {
+
+struct ArrayInState {
```
Comment (amorynan) [inline be/src/vec/functions/function_array_in.h:52]: code is duplicated

**Comment by amorynan** [comment]
just keep a function_collection_in.h/cpp file to handle array/struct/map this collection type.

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

**Comment by taptao** [comment]
run buildall

File: be/src/vec/functions/function_array_in.h (hunk 1 above)
Comment (taptao) [reply be/src/vec/functions/function_array_in.h]: done！

File: be/src/vec/functions/function_collection_in.h [hunk 2]
```diff
@@ -98,14 +98,17 @@ class FunctionStructIn : public IFunction {
...
         context->set_function_state(scope, state);
-        DCHECK(context->get_num_args() >= 1);
+
         auto* col_desc = context->get_arg_type(0);
-        DataTypePtr args_type = DataTypeFactory::instance().create_data_type(*col_desc);
-        MutableColumnPtr column_struct_ptr_args = remove_nullable(args_type)->create_column();
-        NullMap null_map(context->get_num_args(), false);
-        for (int i = 1; i < context->get_num_args(); ++i) {
+        DataTypePtr args_type = DataTypeFactory::instance().create_data_type(*col_desc, false);
+        MutableColumnPtr args_column_ptr = remove_nullable(args_type)->create_column();
```
Comment (amorynan) [inline be/src/vec/functions/function_collection_in.h:109]: create_data_type(*col_desc, false); u set false already , there is no need to remove_nullable again

File: be/src/vec/functions/function_collection_in.h [hunk 3]
```diff
@@ -115,36 +118,35 @@ class FunctionStructIn : public IFunction {
             const auto& [col, _] = unpack_if_const(const_column_ptr->column_ptr);
             if (col->is_nullable()) {
                 auto* null_col = vectorized::check_and_get_column<vectorized::ColumnNullable>(col);
-                if (null_col->has_null()) {
+                if (!state->null_in_set && null_col->has_null()) {
```
Comment (amorynan) [inline be/src/vec/functions/function_collection_in.h:121]: this if else logic maybe not right ! 
think about this in params: (null, xxx, null), the second null will be inserted into args_column_ptr

File: be/src/vec/functions/function_collection_in.h [hunk 4]
```diff
@@ -155,18 +157,25 @@ class FunctionStructIn : public IFunction {
...
                 vec_res[i] = !find;
             } else {
                 vec_res[i] = find;
             }
-            if (in_state->null_in_set) {
-                vec_null_map_to[i] = negative == vec_res[i];
-            } else {
-                vec_null_map_to[i] = false;
+
+            if (!find && null_in_set) {
```
Comment (amorynan) [inline be/src/vec/functions/function_collection_in.h:177]: do  not make if not have else as possible as u can ! make code clear for read . and as less as condition in if , the compiler can make more optimize behavior.

**Comment by taptao** [comment]
run buildall

**Review by amorynan — APPROVED** [review]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**